### PR TITLE
feat: allow multi app-ids and auths for fcm

### LINKED ISSUE #1340: Make FCM support multiple senderids
Clients want to use different project ids for each browser. This means that we need to support multiple FCM ids (like we do with multiple GCM ids).

### PR DESCRIPTION
**Pull request by jrconlin**
BREAKING CHANGE:
This now supports multiple registrations for FCM. See the updated docs
and sample configurations.

Obsolete configuration options:
    * fcm_auth
    * fcm_senderid
    * fcm_projectid
    * fcm_service_cred_path

New Configuration
```json
fcm_creds={"_profile_": {"projectid": "_fcm_projectid_", "auth":
"_fcm_service_cred_path_"}}
```

`v1` protocol is now the default.

for existing FCM, _profile_ == _fcm_projectid_

Closes #1340

### CODE REVIEW

File: autopush/config.py [hunk 1]
```diff
@@ -283,28 +283,29 @@ def from_argparse(cls, ns, **kwargs):
...
-            if len(ns.fcm_auth) > 0:
-                if not ns.fcm_senderid:
-                    raise InvalidConfig("No SenderID found for FCM")
-                fcm_core.update({
-                    "version": 0,
-                    "auth": ns.fcm_auth,
-                    "senderID": ns.fcm_senderid})
-            if len(ns.fcm_service_cred_path) > 0:
-                fcm_core.update({
+        if ns.fcm_enabled and ns.fcm_creds:
```
Comment (pjenvey) [inline autopush/config.py]: `fcm_enabled` is now deprecated though?

File: autopush/main_argparse.py [hunk 2]
```diff
@@ -141,6 +141,11 @@ def _obsolete_args(parser):
     parser.add_argument('--storage_read_throughput', help="OBSOLETE")
     parser.add_argument('--storage_write_throughput', help="OBSOLETE")
 
+    parser.add_argument('--fcm_enabled', help="OBSOLETE")
+    parser.add_argument('--fcm_project_id', help="OBSOLETE")
+    parser.add_argument('--fcm_service_cred_path', help="OBSOLETE")
+    parser.add_argument('--fcm_senderid_list', help="OBSOLETE")
```
Comment (pjenvey) [inline autopush/main_argparse.py]:
```suggestion
    parser.add_argument('--fcm_senderid', help="OBSOLETE")
```

File: autopush/main_argparse.py [hunk 3]
```diff
@@ -141,6 +141,11 @@ def _obsolete_args(parser):
     parser.add_argument('--storage_read_throughput', help="OBSOLETE")
     parser.add_argument('--storage_write_throughput', help="OBSOLETE")
 
+    parser.add_argument('--fcm_enabled', help="OBSOLETE")
```
Comment (pjenvey) [inline autopush/main_argparse.py]: (re above) or was this the intention?

```suggestion
    parser.add_argument('--fcm_auth', help="OBSOLETE")
```

File: autopush/tests/test_z_main.py [hunk 4]
```diff
@@ -385,6 +384,7 @@ def test_memusage(self):
         ], False, resource=autopush.tests.boto_resource)
 
     def test_client_certs_parse(self):
+        print(self.TestArg.fcm_creds)
```
Comment (pjenvey) [inline autopush/tests/test_z_main.py]: kill

File: autopush/config.py (hunk 1 above)
Comment (jrconlin) [reply autopush/config.py]: thanks, forgot to get that. (TIL `parser.add_argument` still sets boolean state if the arg exists and no other options set.)

File: autopush/router/fcm_v1.py [hunk 5]
```diff
@@ -26,34 +38,33 @@ def __init__(self, conf, router_conf, metrics):
...
-        self.senderID = router_conf.get("senderID")
         self.version = router_conf["version"]
         self.log = Logger()
-        self.fcm = FCMv1(project_id=self.senderID,
-                         service_cred_path=router_conf['service_cred_path'],
-                         logger=self.log,
-                         metrics=self.metrics)
+        self.clients = {}
+        try:
+            map(self._setter, router_conf["creds"].items())
```
Comment (pjenvey) [inline autopush/router/fcm_v1.py]: ditto

File: autopush/router/fcm.py [hunk 6]
```diff
@@ -109,42 +117,37 @@ def __init__(self, conf, router_conf, metrics):
...
-        self.senderID = router_conf.get("senderID")
-        self.auth = router_conf.get("auth")
-        self._base_tags = ["platform:fcm"]
+        self.clients = {}
         try:
-            self.fcm = pyfcm.FCMNotification(api_key=self.auth)
-        except Exception as e:
-            self.log.error("Could not instantiate FCM {ex}",
-                           ex=e)
+            map(self._setter, router_conf["creds"].items())
```
Comment (pjenvey) [inline autopush/router/fcm.py]: nit: for loops are preferred when not using the list returned by map

File: docs/running.rst [hunk 7]
```diff
@@ -176,25 +176,34 @@ images, that you ***always*** supply a `CRYPTO_KEY` as part of the run command.
...
+
+   * add ``fcm_enabled`` to enable FCM routing.
+
+   * add ``fcm_creds``. This is a json block with the following format:
+
+     {"**app id**": {"projectid": "**project id name**", "auth": "**path to Private Key File**"}, ...}
+
+where:
+
+**profile_name**: the URL identifier to be used when registering endpoints. (e.g. if "reference_test" is
```
Comment (pjenvey) [inline docs/running.rst]:
```suggestion
**app id**: the URL identifier to be used when registering endpoints. (e.g. if "reference_test" is
```

**Pull request merged**